pkg/util/network.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"net"
+	"regexp"
+)
+
+// hostNameFormat matches strings that:
+//
+//   - do not exceed 63 characters
+//   - contain only alpha-numeric characters, dashes, or symbol-point unicode
+//     (ex. emoji)
+//   - does not begin or end with a dash
+const hostNameFormat = `^([a-zA-Z0-9\p{S}\p{L}]{1,2}|(?:[a-zA-Z0-9\p{S}\p{L}][a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}]))$`
+
+var hostNameRx = regexp.MustCompile(hostNameFormat)
+
+// IsValidHostName returns true if the provided value adheres to the format of
+// an RFC-1123 DNS label, which may be:
+//
+//   - an IP4 address
+//   - an IP6 address
+//   - a string of alpha-numeric characters that:
+//   - must not exceed 63 characters in length
+//   - may begin with a digit
+//   - may contain dashes except as the first or last character
+//   - must not contain underscores
+//   - may contain symbol-point unicode, (ex. emoji)
+func IsValidHostName(s string) bool {
+	return net.ParseIP(s) != nil || hostNameRx.MatchString(s)
+}
+
+// fqdnFormat matches strings that are fully-qualified domain names.
+const fqdnFormat = `^((?:(?:(?:[a-zA-Z0-9\p{S}\p{L}]{1,2})|(?:[a-zA-Z0-9\p{S}\p{L}][-a-zA-Z0-9\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}]))\.){0,}(?:(?:[a-zA-Z0-9]{2})|(?:[a-zA-Z0-9][-a-zA-Z0-9]{0,61}[a-zA-Z0-9])))$`
+
+var fqdnNameRx = regexp.MustCompile(fqdnFormat)
+
+// IsValidDomainName returns true if the provided value adheres to the format
+// for DNS names specified in RFC-1034, Section 3.5:
+//
+//   - must not exceed 255 characters in length when combined with the host name
+//   - individual segments must be 63 characters or less.
+//   - the top-level domain( ex. ".com"), is at least two letters with no
+//     special characters.
+//   - underscores are not allowed.
+//   - dashes are permitted, but not at the start or end of the value.
+//   - long, top-level domain names (ex. ".london") are permitted.
+//   - symbol unicode points, such as emoji, are disallowed in the top-level
+//     domain.
+func IsValidDomainName(s string) bool {
```

Comment:
Similar comment.  Maybe use the upstream validation? https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L95

Reply:
Nope, see https://github.com/vmware-tanzu/vm-operator/pull/487#discussion_r1585336668.